File: src/main/java/me/nobaboy/nobaaddons/features/chatcommands/ChatCommandManager.java
package me.nobaboy.nobaaddons.features.chatcommands;

import jline.internal.Nullable;
import me.nobaboy.nobaaddons.NobaAddons;
import me.nobaboy.nobaaddons.util.CooldownManager;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;

public abstract class ChatCommandManager extends CooldownManager {
    private final List<IChatCommand> commands = new ArrayList<>();
    private final Object $lock = new Object();

    protected void register(IChatCommand command) {
        this.commands.add(command);
    }

    public List<IChatCommand> getCommands() {
        return commands;
    }

    /**
     * Match the provided message with a regex pattern for a possible chat command
     *
     * @param message The message received without any formatting codes
     * @return An {@link Optional} containing a {@link Matcher} with the named groups {@code username}, {@code command}, and {@code argument}
     */
    protected abstract Optional<Matcher> matchMessage(String message);

    public Optional<ChatContext> getContext(String message) {
        return matchMessage(message).map(match -> {
            String user = match.group("username");
            String command = match.group("command");
            @Nullable String args = match.group("argument");
            return new ChatContext(user, command, args != null ? args.split(" ") : new String[]{}, message);
        });
    }

    @SuppressWarnings("CodeBlock2Expr")
    public void processMessage(String message) {
        synchronized ($lock) {
            if(isOnCooldown()) return;

            getContext(message).ifPresent(ctx -> {
                commands.stream()
                        .filter(IChatCommand::isEnabled)
                        .filter(x -> x.name().equalsIgnoreCase(ctx.command()) ||
                                x.aliases().stream().anyMatch(alias -> alias.equalsIgnoreCase(ctx.command())))
                        .findFirst()
                        .ifPresent(cmd -> {
                            try {
                                cmd.run(ctx);
                                if(!NobaAddons.config.debugMode) {
                                    startCooldown();
                                }
                            } catch(Exception e) {
                                // TODO proper error feedback
                                e.printStackTrace();
                            }
                        });
            });
        }
    }
}

File: src/main/kotlin/me/nobaboy/nobaaddons/features/chatcommands/ChatCommandManager.kt
package me.nobaboy.nobaaddons.features.chatcommands

import me.nobaboy.nobaaddons.NobaAddons
import me.nobaboy.nobaaddons.util.CooldownManager
import me.nobaboy.nobaaddons.util.StringUtils.lowercaseEquals
import java.util.regex.Matcher

abstract class ChatCommandManager : CooldownManager() {
    private val commands = mutableListOf<IChatCommand>()
    private val lock = Object()

    protected fun register(command: IChatCommand) {
        commands.add(command)
    }

    fun getCommands(): List<IChatCommand> {
        return commands
    }

    protected abstract fun matchMessage(message: String): Matcher?

    private fun getContext(message: String): ChatContext? {
        val match = matchMessage(message) ?: return null
        val user = match.group("username")
        val command = match.group("command")
        val args = match.group("argument")?.split(" ")?.toList() ?: emptyList()
        return ChatContext(user, command, args, message)
    }

    fun processMessage(message: String) {
        synchronized(lock) {
            val ctx = getContext(message) ?: return
            val cmd = commands.asSequence()
                .filter { it.isEnabled }
                .firstOrNull {
                    it.name.lowercaseEquals(ctx.command()) ||
                    it.aliases.any { alias ->
                        alias.lowercaseEquals(ctx.command())
                    }
                } ?: return

            if (!cmd.bypassCooldown && isOnCooldown()) return
            try {
                cmd.run(ctx)
                if (!NobaAddons.config.debugMode && !cmd.bypassCooldown) startCooldown()
            } catch(ex: Exception) {
                NobaAddons.LOGGER.error("Failed to run chat command '$cmd' | Name: '${cmd.name}`.", ex)
            }
        }
    }
}